(DOCSP-26938) [Section] Define a Data Model (#350)
mongodb/docs-app-services: source/data-model.txt

.. _data-model:

===================
Define a Data Model
===================

.. contents:: On this page
   :depth: 1

.. toctree::

   Schemas </schemas>
   Relationships </schemas/relationships>
   Values & Secrets </values-and-secrets>

Overview
--------

Your App's data model is a description of the data that your App uses.

> Your data model helps to ensure that your data is well-formed, contains

Reviewer: Nit: the data model doesn't actually ensure that the data is "well-formed." If the data model isn't well-formed, the data won't be well-formed. It's probably more accurate to say that something about how it enforces a specific data structure.

```suggestion
Your data model ensures that your data complies with a specific structure, contains
```

only valid values, and only changes in ways that you expect.

> A data model is composed of several components, including:

Reviewer: Nit: passive voice.
```suggestion
A data model consists of several components, including:
```

- A set of relationships that define logical connections between your
  schema.

- A set of static values that you can use to define global constants,
  environment values, and secret values.

- A set of rule-based query filters that you can use to dynamically
  limit MongoDB operations to a subset of all data for a given request.

Defining your App's data model is a key part of configuring and using
Device Sync. You don't need to define a data model to use other App
Services. However, a data model unlocks many powerful features that can
save you time and let you focus on building features instead of writing
boilerplate code.

For example, once you define a data model, you can:

- Automatically generate a GraphQL schema with rich types, queries, and
  mutations for every document schema. To learn more, see :ref:`GraphQL
  API <graphql-api>`.

- Automatically generate syncable Realm object models in your native
  programming language using an App administration tool. To learn more,
  see :ref:`Generate SDK Object Models <generate-sdk-object-models>`.

Document Schemas
----------------

A document schema is a JSON object that defines the structure of a

> particular type of data that your App uses. Document schemas control

Reviewer: This is just a nit, but pedantically, I want something else here because a schema doesn't directly do the controlling/enforcing. It's just a blueprint, which our very complicated rules engine uses to enforce these things. 

```suggestion
particular type of data that your App uses. Document schemas describe
```

which fields a document has, the type of values those fields contain,
and conditions that must be met for a change of value to be valid.

You define schemas at the collection level. This means that you
associate each schema with a particular MongoDB collection that holds
the documents of that type. Schemas use a superset of the JSON schema
standard that includes support for MongoDB's built-in BSON types, which
allows you to fully describe your data in MongoDB.

Your App enforces schemas at runtime whenever data is written to
MongoDB. This includes inserts, updates, and deletes from an API
service, Function, or Device Sync.

To learn more, see :ref:`Schemas <schemas>`.

Relationships
-------------

A relationship is a connection between two document schemas that lets

> you logically associate one document with one or more other documents.

Reviewer: Couldn't relationships technically have zero other documents? i.e. an empty list or optional to-one relationship?

I wonder if we should instead say "document type"/ "document types" in which case this would make sense?

```suggestion
you logically associate one document with zero or more other documents.
```

In a relationship, a "source" document stores a reference to related
documents, for example a list of related document ``_id`` values. When
you use Device Sync or the GraphQL API, App Services automatically
resolves the references to the related documents so you can query them
directly.

To learn more, see :ref:`Relationships <relationships>`.

Values & Secrets
----------------

A value is a named constant that you can access by name in various ways
throughout your App. For example, you can use a value to define the base
URL of an external API service and then reference the value instead of
duplicating the base URL across many Function definitions.

You can also define an environment value, which can change depending on

> your App's environment tag. For example, you might use a different API

Reviewer: Should this link to the "environment" documentation? (And should above link to the "Function" documentation? I don't know where this lands on the spectrum of "unnecessary wiki-style link" vs "easily give users a link to related docs so they can understand the context."

Author: I don't think it's necessarily too much linking, will add
